### PR TITLE
Document the nondeterministic properties of `AStar.get_closest_point()`

### LINKED ISSUE #3667: AStar (& AStar2D) get_closest_point() needs additional clarifying information.
**Your Godot version:**
Godot 3.2.1.stable

**Issue description:**
AStar (& AStar2D) Docs need to be updated with clarifying information.

When 2 different AStar IDs share the same cellv position, get_closest_point() appears to return one of them at random. (e.g. ID 200 & ID 250, with the same position, either one may be returned.)

If this is indeed the case, we should mention in the Docs that get_closest_point() does not guarantee a consistent, specific astar ID will be returned (e.g. the smaller ID).

**URL to the documentation page:**
https://docs.godotengine.org/en/stable/classes/class_astar.html#class-astar-method-get-closest-point
https://docs.godotengine.org/en/stable/classes/class_astar2d.html#class-astar2d-method-get-closest-point

### PR DESCRIPTION
**Pull request by Calinou**
This closes https://github.com/godotengine/godot-docs/issues/3667.

### CODE REVIEW

File: doc/classes/AStar.xml [hunk 1]
```diff
@@ -131,7 +131,8 @@
 			<argument index="1" name="include_disabled" type="bool" default="false">
 			</argument>
 			<description>
-				Returns the ID of the closest point to [code]to_position[/code], optionally taking disabled points into account. Returns -1 if there are no points in the points pool.
+				Returns the ID of the closest point to [code]to_position[/code], optionally taking disabled points into account. Returns [code]-1[/code] if there are no points in the points pool.
+				[b]Note:[/b] If there are several closest points at the same position, this method will return one of their IDs in a seemingly random manner.
```
Comment (akien-mga) [inline doc/classes/AStar.xml:135]: I would say "non-deterministic" instead of "seemingly random".

**Comment by akien-mga** [comment]
Alternatively, we could make it deterministic with this (untested, but should work I guess):
```diff


        diff --git a/core/math/a_star.cpp b/core/math/a_star.cpp
        index 580a7cf7bb..576f180608 100644
        --- a/core/math/a_star.cpp
        +++ b/core/math/a_star.cpp
        @@ -280,10 +280,12 @@ int AStar::get_closest_point(const Vector3 &p_point, bool p_include_disabled) co
                                continue; // Disabled points should not be considered.
                        }
         
        +               // Keep the closest ID, and in case of equality, the smallest (makes it deterministic).
                        real_t d = p_point.distance_squared_to((*it.value)->pos);
        -               if (closest_id < 0 || d < closest_dist) {
        +               int id = *(it.key);
        +               if (closest_id < 0 || d < closest_dist || (d == closest_dist && id < closest_id)) {
                                closest_dist = d;
        -                       closest_id = *(it.key);
        +                       closest_id = id;
                        }
                }
```
But not sure if we consider it worth the cost of that extra check?

**Comment by Calinou** [comment]
I think making it deterministic would be better indeed. If it can be done without losing too much performance, we should go for that instead.

Can someone try to benchmark the performance difference? cc @eltharionx

**Comment by akien-mga** [comment]
I'll make a PR, and hopefully someone can benchmark it.

**Comment by akien-mga** [comment]
Superseded by #39409.

**Pull request closed without merging**